README.md
# NewsAggregator
Backend using FLASK and PostgreSQL
### Given Tasks as follows :
Given two csv files. Following data is present in the two files.
- The `stock_industry_mapping.csv` has a all stocks and industry that are present in india.
    1. id → Unique id for stock
    2. ticker → name of the stocks’s ticker *******************************ticker is a symbol for the stock*******************************
    3. stock → name of the stock
    4. industry → name of the industry.
- 2. the `news.csv` has multiple news where news are mapped to stocks, for news that is not related to any stock the value is filled with `nan`
    1. id → unique id for news
    2. title → heading of the news
    3. link → URL for the news
    4. summary → text summary describing the news
    5. published → when was the news created
    6. source → news source
    7. org → if the news is related to an organisation
    8. mapped_stock_id → stock id from the csv `stock_industry_mapping.csv`
    9. ticker → ticker of the mapped stock
    10. stock → name of the mapped stock
    11. industry → industry of the mapped stock

Given these two files perform the following objectives
1. Create Two different endpoints to push the data from the two csv in separate tables inside the database. You are free to use any database. Do justify the reason of choosing the database
2. A user should be able to perform the following tasks using API/Endpoints.
    1. Get all the list of industries, stock, ticker that are present in India.
    2. Given an industry get all the stocks along with the tickers. *(Used the get method here which passes parameters through the url to get desired data)* 
    3. Get all news at ones 
    4. Get news by news-id (News id is casted to its unique UUID format)
    5. Get all news related to a ticker  *(Used the get method here which passes multiple ticker parameters through the url to get desired data)*
    6. Get all news related to an industry *(Used the get method here which passes multiple industry names as parameters through the url to get desired data)* 
    7. Get all news related to a source *(Used the get method here which passes multiple source names as parameters through the url to get desired data)*

### The Deployed on Railway and tested using postman

   deployment website in description

### <a href="https://github.com/OmkarBorker/NewsAggregator/tree/main/tests">The Tests conducted are stored in test directory</a>

## The Hierarchy of the flask app are
    app.py
    ├── __init__.py
        ├── main.py
        └── data.py
- The app runs through app.py 
- The app is setup through __init__.py
- main.py contains all the user functions as mentioned in part 2
- data.py contains all the post operations through which data is posted onto the system 

## The Database and App functionality
- PostgresSQL as the database for the following reasons:
    - PostgreSQL is a powerful open-source RDBMS with strong ACID compliance, making it suitable for data integrity and reliability.
    - It supports complex queries, indexing, and advanced features like JSONB and UUID data type, which can be useful for handling diverse data.
    - It has good integration capabilities with Python and Flask especially with psycopg2 library.
    - As the Schema for the database is given to use it is much prefered to make a structured database using postgresSQL for data analysis and         manipulation, making it easy for isolation of specific data.
- The Fuctionality of the app includes 
    - > _/industries_  
    - Get all the list of industries, stock, ticker that are present in India
    - > _stocks-by-industry?industry=industry1&industry=industry2_
    -  Given an industry get all the stocks along with the tickers ( parameter passing         through url takes care of multiple inputs through GET request)    <p></p>
    - > _/news_ 
    - Get all news at ones
    - > _/news/<id<id>>_ 
    - Get news by news-id
    - > _/news-by-ticker?ticker=ticker1&ticker=ticker2_ 
    - Get all news related to a ticker (parameter passing through url takes care of multiple inputs GET request)
    - > _/news-by-industry?industry=industry1&industry=industry2_ 
    - Get all news related to an industry 
    - > _/news-by-source?source=source1&source=source2_ 
    - Get all news related to a source
    
    <p></p>
    
    -  - **POST request**
    - > _/push-stock-industry-mapping_ 
    - data of industries into table stock_industry_mappings
    - > _/push-news_
    - data of news into table news
    - both endpoints are different to push the data from the two csv in separate tables inside the database.



tests/news.txt
https://newsaggregator-production.up.railway.app/news

{
    "data": [
        [
            "Indoco Remedies receives EU GMP certification for its Himachal Pradesh manufacturing facility",
            "https://www.cnbctv18.com/business/companies/indoco-remedies-eu-gmp-certification-baddi-himachal-pradesh-solan-oral-dosage-facility-16894121.htm",
            "The inspection was conducted by the German Health Authority at the companyâ€™s manufacturing unit from May 22-25, 2023.",
            "Fri, 09 Jun 2023 16:17:01 GMT",
            "cnbc",
            "Indoco Remedies",
            "INDOCO",
            "Indoco Remedies Ltd.",
            "Pharmaceuticals & Drugs",
            "118788d8-25b6-11ec-afd7-8c859079d2ba",
            "2d903f26-0885-11ee-ad77-acde48001122"
        ],
        [
            "Macquarie downgrades HDFC Life, ICICI Prudential, SBI Life to neutral, cuts target prices",
            "https://www.cnbctv18.com/finance/macquarie-downgrades-hdfc-life-icici-prudential-sbi-life-to-neutral-cuts-target-prices-16890921.htm",
            "While the target price for HDFC Life Insurance has been cut to Rs 556 from Rs 655 earlier, ICICI Prudential Life target price has been slashed Rs 486 from Rs 580 and SBI Life Insurance target price to Rs 1,306 from Rs 1,580 a share.",
            "Fri, 09 Jun 2023 13:52:05 GMT",
            "cnbc",
            "HDFC Life",
            "HDFCLIFE",
            "HDFC Life Insurance Co Ltd.",
            "Insurance",
            "1187f6f6-25b6-11ec-966b-8c859079d2ba",
            "2d903ff8-0885-11ee-ad77-acde48001122"
        ],
        [
            "Macquarie downgrades HDFC Life, ICICI Prudential, SBI Life to neutral, cuts target prices",
            "https://www.cnbctv18.com/finance/macquarie-downgrades-hdfc-life-icici-prudential-sbi-life-to-neutral-cuts-target-prices-16890921.htm",
            "While the target price for HDFC Life Insurance has been cut to Rs 556 from Rs 655 earlier, ICICI Prudential Life target price has been slashed Rs 486 from Rs 580 and SBI Life Insurance target price to Rs 1,306 from Rs 1,580 a share.",
            "Fri, 09 Jun 2023 13:52:05 GMT",
            "cnbc",
            "ICICI Prudential",
            "nan",
            "nan",
            "nan",
            null,
            "2d903ff8-0885-11ee-ad77-acde48001122"
        ],
        [
            "Macquarie downgrades HDFC Life, ICICI Prudential, SBI Life to neutral, cuts target prices",
            "https://www.cnbctv18.com/finance/macquarie-downgrades-hdfc-life-icici-prudential-sbi-life-to-neutral-cuts-target-prices-16890921.htm",
            "While the target price for HDFC Life Insurance has been cut to Rs 556 from Rs 655 earlier, ICICI Prudential Life target price has been slashed Rs 486 from Rs 580 and SBI Life Insurance target price to Rs 1,306 from Rs 1,580 a share.",
            "Fri, 09 Jun 2023 13:52:05 GMT",
            "cnbc",
            "Macquarie",
            "nan",
            "nan",
            "nan",
            null,
            "2d903ff8-0885-11ee-ad77-acde48001122"
        ],
        [
            "Macquarie downgrades HDFC Life, ICICI Prudential, SBI Life to neutral, cuts target prices",
            "https://www.cnbctv18.com/finance/macquarie-downgrades-hdfc-life-icici-prudential-sbi-life-to-neutral-cuts-target-prices-16890921.htm",
            "While the target price for HDFC Life Insurance has been cut to Rs 556 from Rs 655 earlier, ICICI Prudential Life target price has been slashed Rs 486 from Rs 580 and SBI Life Insurance target price to Rs 1,306 from Rs 1,580 a share.",
            "Fri, 09 Jun 2023 13:52:05 GMT",
            "cnbc",
            "SBI Life",
            "SBILIFE",
            "SBI Life Insurance Company Ltd.",
            "Insurance",
            "1187ea26-25b6-11ec-897e-8c859079d2ba",
            "2d903ff8-0885-11ee-ad77-acde48001122"
        ],
        [
            "Zydus Lifesciences begins Phase-IV trial for non-alcoholic fatty liver disease drug Â ",
            "https://www.cnbctv18.com/healthcare/zydus-lifesciences-phase-iv-trial-non-alcoholic-fatty-liver-disease-drug-nfld-comorbidities-16892151.htm",
            "This will include a total of around 1,500 male and female NAFLD patients having comorbidities like obesity, dyslipidemia, type 2 diabetes mellitus or metabolic syndrome.",
            "Fri, 09 Jun 2023 13:56:48 GMT",
            "cnbc",
            "Zydus Lifesciences",
            "SMSLIFE",
            "SMS Lifesciences India Ltd.",
            "Pharmaceuticals & Drugs",
            "11898bb0-25b6-11ec-8e40-8c859079d2ba",
            "2d90400c-0885-11ee-ad77-acde48001122"
        ],
        [
            "Asia stocks advance on hells of S&P 500â€™s bull run",
            "https://www.cnbctv18.com/market/asia-stocks-advance-on-hells-of-sp-500s-bull-run-16888601.htm",
            "The rise for Japanese equities snapped a two-day decline while Australian shares rebounded after falling in the prior three sessions. The improvements, combined with an advance for Hong Kong futures, place a regional equity index on pace for its second weekly gain.",
            "Fri, 09 Jun 2023 06:23:31 GMT",
            "cnbc",
            "S & P",
            "nan",
            "nan",
            "nan",
            null,
            "2d9041ba-0885-11ee-ad77-acde48001122"
        ],
        [
            "Large deal likely in Kotak Bank, MedPlus, HNIs buy Shree Renuka - Stocks that kept dealers busy on Thursday",
            "https://www.cnbctv18.com/market/dealing-room-check-kotak-bank-medplus-shree-renuka-religare-nifty-50-16884741.htm",
            "Metal stocks and power-related stocks were well bid, according to dealers. Power Finance Corporation ended higher for the ninth day in a row.",
            "Thu, 08 Jun 2023 16:19:41 GMT",
            "cnbc",
            "HNI",
            "nan",
            "nan",
            "nan",
            null,
            "2d90421e-0885-11ee-ad77-acde48001122"
        ],

main/data.py
import csv
from flask import Blueprint, request, current_app, jsonify
import psycopg2

DB_URL = 'postgres://postgres:Omkarborker1@db.jjdoiyhcuilwxtdkpbml.supabase.co:6543/postgres'

def create_db_connection():
    conn = psycopg2.connect(DB_URL)
    return conn

data_bp = Blueprint('data', __name__)

@data_bp.route('/', methods=['GET'])
def test_db_connection():
    try:
        conn = create_db_connection()
        conn.close()
        return jsonify({'message': 'Connected to the PostgreSQL database successfully'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@data_bp.route('/push-stock-industry-mapping', methods=['POST'])
def push_stock_industry_mapping():
    try:
        conn = create_db_connection()
        cursor = conn.cursor()
        cursor.execute('TRUNCATE TABLE stock_industry_mapping;')

        with open('stock_industry_mapping.csv', 'r') as file:
            csv_data = csv.reader(file)
            next(csv_data)  # Skip header row

            for row in csv_data:
                query = 'INSERT INTO stock_industry_mapping (id, ticker, stock, industry) VALUES (%s, %s, %s, %s);'
                cursor.execute(query, row)

        conn.commit()
        cursor.close()
        conn.close()
        return jsonify({'message': 'Data inserted into stock_industry_mapping table successfully'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@data_bp.route('/push-news', methods=['POST'])
def push_news():
    try:
        conn = create_db_connection()
        cursor = conn.cursor()
        cursor.execute('TRUNCATE TABLE news;')

        with open('news.csv', 'r') as file:
            csv_data = csv.reader(file)
            next(csv_data)  # Skip header row

            for row in csv_data:
                if row[7] == '':
                    row[7] = None
                query = 'INSERT INTO news (id, title, link, summary, published, source, org, mapped_stock_id,ticker, stock, industry ) VALUES (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s);'
                cursor.execute(query, row)

        conn.commit()
        cursor.close()
        conn.close()
        return jsonify({'message': 'Data inserted into news table successfully'})
    except Exception as e:
        return jsonify({'error': str(e)}), 500


main.py
from flask import jsonify,request,Blueprint, current_app
import psycopg2
import urllib.parse


DB_URL = 'postgres://postgres:Omkarborker1@db.jjdoiyhcuilwxtdkpbml.supabase.co:6543/postgres'

def create_db_connection():
    conn = psycopg2.connect(DB_URL)
    return conn

main_bp = Blueprint('main', __name__)

@main_bp.route('/industries', methods=['GET'])
def get_industries():
    try:
        
        conn = create_db_connection()
        cursor = conn.cursor()

        # Execute query to get list of industries, stocks, and tickers in India
        query = "SELECT industry, stock, ticker FROM stock_industry_mapping"
        cursor.execute(query)
        result = cursor.fetchall()

        cursor.close()
        return {'data': result}
    except Exception as e:
        return {'error': str(e)}

@main_bp.route('/stocks-by-industry', methods=['GET'])
def get_stocks_by_industry():
    try:
        industries = request.args.getlist('industry')

        if not industries:
            return jsonify({'error': 'No industries provided'}), 400

        conn = current_app.config['DB_CONNECTION']
        cursor = conn.cursor()

        # Construct the SQL query to retrieve stocks and tickers for the given industries
        query = "SELECT stock, ticker FROM stock_industry_mapping WHERE industry IN %s"
        cursor.execute(query, (tuple(industries),))  # Pass the industries as a tuple

        result = cursor.fetchall()
        cursor.close()

        return jsonify({'data': result})
    except Exception as e:
        return jsonify({'error': str(e)}), 500

@main_bp.route('/news', methods=['GET'])
def get_all_news():
    try:
        conn = create_db_connection()
        cursor = conn.cursor()

        # Execute query to get all news items
        query = "SELECT * FROM news"
        cursor.execute(query)
        result = cursor.fetchall()

        cursor.close()
        return {'data': result}
    except Exception as e:
        return {'error': str(e)}

@main_bp.route('/news/<string:news_id>', methods=['GET'])
def get_news_by_id(news_id):
    try:
        conn = create_db_connection()
        cursor = conn.cursor()
        # Execute query to get news details for the given news ID
        query = " SELECT * FROM news WHERE id = %s::uuid "
        cursor.execute(query, (news_id,))
        result = cursor.fetchone()

        cursor.close()
        return {'data': result}
    except Exception as e:
        return {'error': str(e)}

@main_bp.route('/news-by-ticker', methods=['GET'])
def get_news_by_ticker():
    try:
        tickers = request.args.getlist('ticker')  # Get the list of tickers from the query parameters

        conn = create_db_connection()
        cursor = conn.cursor()

        # Construct the SQL query to retrieve news items for the given tickers
        query = "SELECT * FROM news WHERE ticker IN %s"
        cursor.execute(query, (tuple(tickers),))  # Pass the tickers as a tuple

        result = cursor.fetchall()
        cursor.close()

        return jsonify({'data': result})
    except Exception as e:
        return jsonify({'error': str(e)}), 500
    
@main_bp.route('/news-by-industry', methods=['GET'])
def get_news_by_industry():
    try:
        industries = request.args.getlist('industry')

        if not industries:
            return jsonify({'error': 'No industries provided'}), 400

        conn = current_app.config['DB_CONNECTION']
        cursor = conn.cursor()

        # Construct the SQL query to retrieve news items for the given industries
        query = "SELECT * FROM news WHERE industry IN %s"
        cursor.execute(query, (tuple(industries),))  # Pass the industries as a tuple

        result = cursor.fetchall()
        cursor.close()

        return jsonify({'data': result})
    except Exception as e:
        return jsonify({'error': str(e)}), 500


@main_bp.route('/news-by-source', methods=['GET'])
def get_news_by_source():
    try:
        sources = request.args.getlist('source')

        if not sources:
            return jsonify({'error': 'No sources provided'}), 400

        conn = current_app.config['DB_CONNECTION']
        cursor = conn.cursor()

        # Construct the SQL query to retrieve news items for the given sources
        query = "SELECT * FROM news WHERE source IN %s"
        cursor.execute(query, (tuple(sources),))  # Pass the sources as a tuple

        result = cursor.fetchall()
        cursor.close()

        return jsonify({'data': result})
    except Exception as e:
        return jsonify({'error': str(e)}), 500
